packages/app/src/elements/audit.ts

```diff
@@ -0,0 +1,239 @@
+                        isWeak = true;
+
+                        // Don't add the same item twice to the list, if there are more than one reused password fields in it
+                        if (!weakPasswordItemIds.has(item.id)) {
+                            weakPasswords.push({ item, vault });
+                            weakPasswordItemIds.add(item.id);
+                        }
+
+                        // TODO: Save audit match boolean in field
+                        // passwordField.auditResult = { ...(passwordField.auditResult || {}), isWeak };
+                    }
+
+                    // TODO: Perform compromised audit
+                }
+
+                // TODO: Save audit match booleans in item
+                // item.auditResult = { lastAudited: new Date(), isReused, isWeak, isCompromised };
+
+                console.log({ item, isReused, isWeak });
+            }
+        }
+
+        // TODO: Sync?
+
+        // This makes the UI update
+        this._reusedPasswords = reusedPasswords;
+        this._weakPasswords = weakPasswords;
+        this._compromisedPasswords = compromisedPasswords;
+
+        this._countElements.forEach((countElement) => {
+            animateElement(countElement, { animation: "bounce" });
+        });
+    }
+
+    private _renderSection(listItems: ListItem[], label: string, routeParams: any) {
+        return html`
+            <section class="box count" ?hidden=${listItems.length === 0}>
+                <h2 class="uppercase bg-dark border-bottom semibold center-aligning spacing horizontal layout">
+                    <div></div>
+                    <div>${label}</div>
+                    <div class="tiny bold tag">${listItems.length}</div>
+                    <div class="stretch"></div>
+                </h2>
+                <pl-list>
+                    ${listItems.slice(0, 5).map(
+                        (listItem) => html`
+                            <div class="list-item hover click" @click=${() => this.go(`items/${listItem.item.id}`)}>
+                                <pl-vault-item-list-item
+                                    .item=${listItem.item}
+                                    .vault=${listItem.vault}
+                                    class="stretch collapse"
+                                ></pl-vault-item-list-item>
+                            </div>
+                        `
+                    )}
+                </pl-list>
+                <pl-button
+                    class="slim margined transparent"
+                    @click=${() => this.go("items", routeParams)}
+                    ?hidden=${listItems.length < 6}
+                >
+                    <div>${$l("Show All")}</div>
+                    <pl-icon icon="arrow-right" class="small left-margined"></pl-icon>
+                </pl-button>
+            </section>
+        `;
+    }
+}
+
+async function sha1(stringToHash: string) {
```

Comment:
I couldn't find this anywhere (and I'll need it for the HIBP as well), and since it isn't exactly "secure", I didn't think it was worth putting it in `core`, but I can be persuaded otherwise.

Reply:
The [crypto](https://github.com/padloc/padloc/blob/master/packages/core/src/crypto.ts) module has got you covered 😉

```ts
import { getCryptoProvider } from "@padloc/core/src/platform";
import { HashParams } from "@padloc/core/src/crypto";
import { stringToBytes } from "@padloc/core/src/encoding";

...

const hashed = await getCryptoProvider()
    .hash(stringToBytes(password), new HashParams({ algorithm: "SHA-1" }));
```